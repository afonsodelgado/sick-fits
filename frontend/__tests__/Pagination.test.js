import 'jest-styled-components'
import { mount } from 'enzyme'
import { MockedProvider } from 'react-apollo/test-utils'
import Router from 'next/router'
import Pagination, { PAGINATION_QUERY } from '../components/Pagination'

Router.router = {
  push: () => null,
  prefetch: () => null
}

const mocks = [
  {
    request: { query: PAGINATION_QUERY },
    result: {
      data: {
        itemsConnection: {
          __typename: 'aggregate',
          aggregate: {
            __typename: 'count',
            count: 12
          }
        }
      }
    }
  }
]

describe('<Pagination />', () => {
  it('should render and show a loading state while fetching data', () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Pagination page={1}/>
      </MockedProvider>
    )
    expect(wrapper.find('Pagination').text()).toContain('Loading...')
  });

  it('should render a pagination for 12 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Pagination page={1}/>
      </MockedProvider>
    )

    await new Promise(resolve => setTimeout(resolve))
    wrapper.update()
    expect(wrapper.find('Pagination').text()).toContain('12 items total')
    expect(wrapper.find('Pagination')).toMatchSnapshot()
  });

  it('should disable prev button and enable next button if on the first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Pagination page={1} />
      </MockedProvider>
    )
    
    await new Promise(resolve => setTimeout(resolve))
    wrapper.update()
    expect(wrapper.find('.prev').props()["aria-disabled"]).toBe(true)
    expect(wrapper.find('.next').props()["aria-disabled"]).toBe(false)
  })

  it('should enable prev button and disable next button if on the last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Pagination page={3} />
      </MockedProvider>
    )
    
    await new Promise(resolve => setTimeout(resolve))
    wrapper.update()
    expect(wrapper.find('.prev').props()["aria-disabled"]).toBe(false)
    expect(wrapper.find('.next').props()["aria-disabled"]).toBe(true)
  })

  it('should enable both prev and next buttons if on a middle page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Pagination page={2} />
      </MockedProvider>
    )
    
    await new Promise(resolve => setTimeout(resolve))
    wrapper.update()
    expect(wrapper.find('.prev').props()["aria-disabled"]).toBe(false)
    expect(wrapper.find('.next').props()["aria-disabled"]).toBe(false)
  })
});