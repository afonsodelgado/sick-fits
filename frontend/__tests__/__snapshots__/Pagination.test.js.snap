// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render a pagination for 12 items 1`] = `
.c0 {
  text-align: center;
  display: inline-grid;
  grid-template-columns: repeat(4,auto);
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  margin: 2rem 0;
  border: 1px solid;
  border-radius: 10px;
}

.c0 > * {
  margin: 0;
  padding: 15px 30px;
  border-right: 1px solid;
}

.c0 > *:last-child {
  border-right: 0;
}

.c0 a[aria-disabled='true'] {
  color: grey;
  pointer-events: none;
}

<Pagination
  page={1}
>
  <Query
    query={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "PAGINATION_QUERY",
            },
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "itemsConnection",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "aggregate",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "count",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 100,
          "start": 0,
        },
      }
    }
  >
    <PaginationStyles>
      <div
        className="c0"
      >
        <SideEffect(Head)>
          <Head />
        </SideEffect(Head)>
        <Link
          href={
            Object {
              "pathname": "/items",
              "query": Object {
                "page": 0,
              },
            }
          }
          prefetch={true}
        >
          <a
            aria-disabled={true}
            className="prev"
            href="/items?page=0"
            onClick={[Function]}
          >
            Prev
          </a>
        </Link>
        <p>
          Page 
          1
           of 
          3
        </p>
        <p>
          12
           items total
        </p>
        <Link
          href={
            Object {
              "pathname": "/items",
              "query": Object {
                "page": 2,
              },
            }
          }
          prefetch={true}
        >
          <a
            aria-disabled={false}
            className="next"
            href="/items?page=2"
            onClick={[Function]}
          >
            Next
          </a>
        </Link>
      </div>
    </PaginationStyles>
  </Query>
</Pagination>
`;
